### ISSUE TITLE
Unexpected conformance violation hit for BanUnknownTypedClassPropsReferences

### ISSUE THREAD

**Issue by supersteves** [google/closure-compiler#1805]
Compiling this with conformance violations enabled:
``` javascript
function x(o) {}

class A {
  constructor() {
    this.elem = null;
  }
  f() {
    x({
      someProp: elem => this.elem = elem 
    });
  }
}
```
results in:
```
input0:10: WARNING - Violation: Unknown typed class props reference
The property "elem" on type "A"
    x({ a: elem => this.elem = elem });
```

I think this is triggered by putting a this.* reference within an object literal, but am not sure.

[See debugger](https://closure-compiler-debugger.appspot.com/#input0%3Dfunction%2520x(o)%2520%257B%257D%250A%250Aclass%2520A%2520%257B%250A%2520%2520constructor()%2520%257B%250A%2520%2520%2520%2520this.elem%2520%253D%2520null%253B%250A%2520%2520%257D%250A%2520%2520f()%2520%257B%250A%2520%2520%2520%2520x(%257B%250A%2520%2520%2520%2520%2520%2520someProp%253A%2520elem%2520%253D%253E%2520this.elem%2520%253D%2520elem%2520%250A%2520%2520%2520%2520%257D)%253B%250A%2520%2520%257D%250A%257D%26input1%26conformanceConfig%3Drequirement%253A%2520%257B%2520type%253A%2520CUSTOM%2520java_class%253A%2520'com.google.javascript.jscomp.ConformanceRules%2524BanUnknownTypedClassPropsReferences'%2520error_message%253A%2520'Unknown%2520typed%2520class%2520props%2520reference'%2520%257D%250A%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26CHECK_TYPES_NEW_INFERENCE%3D1%26CLOSURE_PASS%3D1%26LANG_IN_IS_ES6%3D1%26MISSING_PROPERTIES%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1%26TRANSPILE%3D1).

I came across this when using React + JSX where this...
``` javascript
<Component ref={function...} />
``` 
gets transpiled by Babel to this...
``` javascript
React.createElement(Component, { ref: elem => this.elem = elem })
```
which generates the conformance violation.

**Comment by dimvar** [comment]
In the object literal passed to `x`, the reference to `this` is not typed, so AFAICT this is working as expected.

**Comment by supersteves** [comment]
@dimvar What's strange is it works at runtime after compilation. So why is this. confused within {} but OK at runtime in Chrome?

**Comment by dimvar** [comment]
The compiler's unknown-this warning is conservative. There are definitely cases where it warns and everything is fine at runtime.

**Comment by blickly** [comment]
Hmm, I don't think this has to do with the this, which seems like it gets inferred correctly. I think it's just that it's getting assigned the parameter elem, which is of type unknown.

e.g. With an annotation:
https://closure-compiler-debugger.appspot.com/#input0%3Dfunction%2520x(o)%2520%257B%257D%250A%250Aclass%2520A%2520%257B%250A%2520%2520constructor()%2520%257B%250A%2520%2520%2520%2520this.elem%2520%253D%2520null%253B%250A%2520%2520%257D%250A%2520%2520f()%2520%257B%250A%2520%2520%2520%2520x(%257B%250A%2520%2520%2520%2520%2520%2520%252F**%2520%2540param%2520%257Bnumber%257D%2520elem%2520*%252F%250A%2520%2520%2520%2520%2520%2520someProp%253A%2520elem%2520%253D%253E%2520this.elem%2520%253D%2520elem%2520%250A%2520%2520%2520%2520%257D)%253B%250A%2520%2520%257D%250A%257D%26input1%26conformanceConfig%3Drequirement%253A%2520%257B%2520type%253A%2520CUSTOM%2520java_class%253A%2520'com.google.javascript.jscomp.ConformanceRules%2524BanUnknownTypedClassPropsReferences'%2520error_message%253A%2520'Unknown%2520typed%2520class%2520props%2520reference'%2520%257D%250A%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26CHECK_TYPES_NEW_INFERENCE%3D1%26CLOSURE_PASS%3D1%26LANG_IN_IS_ES6%3D1%26MISSING_PROPERTIES%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1%26TRANSPILE%3D1

**Comment by dimvar** [comment]
@blickly Oh yes, you're right, it's because `elem` is unknown.

@supersteves The error message is confusing, but the conformance check is working as intended, so not much to do here.

**Comment by supersteves** [comment]
Got it. Thanks!
